# Module 2 Report  - Jobs
### Extracting SF Salaries Insights
<br>
By: Billie Kim
<br>
Data Source: https://www.kaggle.com/datasets/san-francisco/sf-employee-compensation

## Contents
1. [Objective](#1.-Objective)
2. [Data Exploration](#2.-Data-Exploration)
3. [Subsets of Data](#3.-Subsets-of-Data)
4. [Single Vector](#4.-Single-Vector)
5. [Five Different Lists](#5.-Five-Different-Lists)
6. [Conditionals](#6.-Conditionals)
7. [Loops](#7.-Loops)
8. [Summary](#8.-Summary)
9. [Recommendations](#9.-Recommendations)

In [1]:
# Load SF Employee Compensation data
df <- read.csv("../mod_report2/data/employee-compensation.csv")

In [3]:
suppressWarnings(library(tidyverse))

## 1. Objective

Examining the cost of living is crucial for maintaining economic stability and improving the well being of individuals residing in communities. By studying income data, the San Francisco Chronicle hopes to give their readers a realistic picture of the current economic and financial factors that affects the workforce in San Francisco.

We are tasked by the newspaper to analyze the cost of living in the city and report insights into income. While two factors affect the cost of living, income and housing, we aim to show how different jobs' income has been changing over the years. We seek to identify trends and across different demographic groups to gain a deeper understanding of income dynamics

This report will use data of workers employed by the city of San Francisco (hosted on Kaggle) to analyze the relationships presented in the data and provide data-driven insights to the newspaper in hopes to educate readers. 

## 2. Data Exploration

In [4]:
# Print number of rows and columns
cat("Number of rows:", nrow(df), "\n")
cat("Number of columns:", ncol(df))

Number of rows: 683277 
Number of columns: 22

**Comment**: We are dealing with a good-sized amount of data with 683,277 records and 22 columns.

In [5]:
# Print top 10 records
head(df, n=10)

Organization.Group.Code Job.Family.Code Job.Code Year.Type Year
1  1                       1000            1021     Calendar  2013
2  1                       1000            1023     Calendar  2013
3  1                       1000            1031     Calendar  2013
4  1                       1000            1054     Calendar  2013
5  1                       1000            1062     Calendar  2013
6  1                       1000            1064     Calendar  2013
7  1                       1000            1070     Calendar  2013
8  1                       1000            1094     Calendar  2013
9  1                       1200            1232     Calendar  2013
10 1                       1600            1654     Calendar  2013
   Organization.Group Department.Code Department          Union.Code
1  Public Protection  ADP             ADP Adult Probation 21        
2  Public Protection  ADP             ADP Adult Probation 21        
3  Public Protection  ADP             ADP Adult Probation 21        
4  Public Protection  ADP             ADP Adult Probation 21        
5  Public Protection  ADP             ADP Adult Probation 21        
6  Public Protection  ADP             ADP Adult Probation 21        
7  Public Protection  ADP             ADP Adult Probation 21        
8  Public Protection  ADP             ADP Adult Probation 21        
9  Public Protection  ADP             ADP Adult Probation 21        
10 Public Protection  ADP             ADP Adult Probation 21        
   Union                                           ... Employee.Identifier
1  Prof & Tech Engineers - Miscellaneous, Local 21 ... 37730              
2  Prof & Tech Engineers - Miscellaneous, Local 21 ... 26574              
3  Prof & Tech Engineers - Miscellaneous, Local 21 ...  8148              
4  Prof & Tech Engineers - Miscellaneous, Local 21 ... 27436              
5  Prof & Tech Engineers - Miscellaneous, Local 21 ... 37730              
6  Prof & Tech Engineers - Miscellaneous, Local 21 ... 36223              
7  Prof & Tech Engineers - Miscellaneous, Local 21 ... 51426              
8  Prof & Tech Engineers - Miscellaneous, Local 21 ... 26574              
9  Prof & Tech Engineers - Miscellaneous, Local 21 ... 37983              
10 Prof & Tech Engineers - Miscellaneous, Local 21 ... 16858              
   Salaries  Overtime Other.Salaries Total.Salary Retirement Health.and.Dental
1   57534.65 0            0.00        57534.65    11253.16   11961.34         
2   57678.50 0            0.00        57678.50    10284.88    7176.80         
3   63532.93 0            0.00        63532.93    12495.83   12918.24         
4  101274.51 0        -7058.59        94215.92    19644.45   12918.24         
5    5084.00 0            0.00         5084.00     1083.90     956.91         
6   34456.00 0            0.00        34456.00     7346.00    3827.63         
7  139889.00 0            0.00       139889.00    26799.23   12918.24         
8   46572.00 0            0.00        46572.00     9929.16    5741.44         
9   84681.53 0            0.00        84681.53    16661.84   12918.24         
10  10611.01 0            0.00        10611.01     1879.20    1435.36         
   Other.Benefits Total.Benefits Total.Compensation
1  4532.63        27747.13        85281.78         
2  4755.14        22216.82        79895.32         
3  4702.71        30116.78        93649.71         
4  7458.76        40021.45       134237.37         
5   387.50         2428.31         7512.31         
6  2579.95        13753.58        48209.58         
7  9791.54        49509.01       189398.01         
8  3822.28        19492.88        66064.88         
9  6838.44        36418.52       121100.05         
10  827.09         4141.65        14752.66

**Comment:** We can't see all of the columns since there are too many so we will print out the column names next.

In [6]:
# Print column names
names(df)

[1] "Organization.Group.Code" "Job.Family.Code"        
 [3] "Job.Code"                "Year.Type"              
 [5] "Year"                    "Organization.Group"     
 [7] "Department.Code"         "Department"             
 [9] "Union.Code"              "Union"                  
[11] "Job.Family"              "Job"                    
[13] "Employee.Identifier"     "Salaries"               
[15] "Overtime"                "Other.Salaries"         
[17] "Total.Salary"            "Retirement"             
[19] "Health.and.Dental"       "Other.Benefits"         
[21] "Total.Benefits"          "Total.Compensation"

In [7]:
# Print column/feature classes
sapply(df, class)

Organization.Group.Code         Job.Family.Code                Job.Code 
              "integer"                "factor"                "factor" 
              Year.Type                    Year      Organization.Group 
               "factor"               "integer"                "factor" 
        Department.Code              Department              Union.Code 
               "factor"                "factor"               "integer" 
                  Union              Job.Family                     Job 
               "factor"                "factor"                "factor" 
    Employee.Identifier                Salaries                Overtime 
              "integer"               "numeric"               "numeric" 
         Other.Salaries            Total.Salary              Retirement 
              "numeric"               "numeric"               "numeric" 
      Health.and.Dental          Other.Benefits          Total.Benefits 
              "numeric"               "numeric"               "numeric" 
     Total.Compensation 
              "numeric"

**Comment:** We seem to have a good amount of both numeric features as well as categorical features. This gives us a pretty decent amount of ways to subset our numerical features based on our categorical features. Going forward, we will be using the Totals such as "Total.Salary, Total.Benefits, and Total.Compensation." 

Next, we will quickly take a look at how many persons our dataset includes by using the Employee.ID field. Then, we will see the years that spans the dataet using the Years field. 

In [8]:
# Count number of unique Employee ID records
length(unique(df$Employee.Identifier))

[1] 309750

In [9]:
# Unique values in the Years column
unique(df$Year)

[1] 2013 2014 2015 2016 2017 2018 2019 2020

**Comment:** Our data spans from 2013 to 2020. According to [sf.gov](https://sf.gov/data/san-francisco-monthly-unemployment#:~:text=As%20of%20July%202022%2C%20the,now%20than%20before%20the%20pandemic.), the labor force in San Francisco is approximately 580,000. Our sample of 309,750 individuals who are employed by the city of San Francisco seems to cover slightly over half of the total work force in the city. It is fair to say that our sample is large enough to make inferences and draw conclusions about the entire population workforce in SF based on the analysis of our sample. 

## 3. Subsets of Data

### 3.1 Salaries by Job Family

In [10]:
# Subset 1
s1.avg.salaries <- df %>% 
    group_by(job_family = Job.Family) %>%
    summarize(avg_salaries = mean(Total.Salary)) %>%
    arrange(desc(avg_salaries))

head(s1.avg.salaries, n=10)

job_family                    avg_salaries
1  Administrative & Mgmt (Unrep) 238194.6    
2  Unassigned                    150105.1    
3  Fire Services                 142523.1    
4  Administrative-Labor & Trades 134588.8    
5  Police Services               128981.3    
6  Construction Project Mgmt     127681.1    
7  Administrative-DPW/PUC        126180.8    
8  Medical & Dental              124301.1    
9  Management                    122460.8    
10 Legal & Court                 115942.6

**Comment:** From this subset, the newspaper can show the average salaries from different job families highlighting variations in salaries in San Francisco.

### 3.2 Top 10 Highest Paying Jobs

In [11]:
# Subset 2
s2.top.jobs <- df %>%
    group_by(Job) %>%
    summarise(total_compensation = mean(Total.Compensation)) %>%
    arrange(desc(total_compensation))

head(s2.top.jobs, n=10)

Job                            total_compensation
1  Chief Investment Officer       496296.9          
2  Physician Administrator, DPH   488861.0          
3  Managing Director              451704.1          
4  Sup Psychiatric Physician Spec 441228.9          
5  Chief, Fire Department         436281.1          
6  Administrator, DPH             418255.7          
7  Director                       400767.1          
8  Assistant Chief of Police      398627.1          
9  Controller                     377978.8          
10 Gen Mgr, Public Trnsp Dept     377815.0

**Comment:** We are able to see which job positions are compensated the most in San Francisco. The information presented in this subset will attract readers and capture their attention.

### 3.3 Trends in Yearly Salaries

In [12]:
# Subset 3
s3.years.salary <- df %>%
    group_by(Year) %>%
    summarise(avg_total_salary = mean(Total.Salary))

head(s3.years.salary, n=10)

Year avg_total_salary
1 2013 72023.35        
2 2014 70218.11        
3 2015 70920.06        
4 2016 74309.36        
5 2017 80290.33        
6 2018 85970.82        
7 2019 86788.81        
8 2020 93996.42

**Comment:** The newspaper can show how salaries have fluctuated over the yeears from 2013-2020. 

### 3.4 Benefits by Departments

In [13]:
# Subset 4
s4.top.benefits <- df %>%
    group_by(Department) %>%
    summarise(avg_retirement = mean(Retirement), avg_health_dental = mean(Health.and.Dental), avg_other_benefits = mean(Other.Benefits), total_benefits = mean(Total.Benefits)) %>%
    arrange(desc(total_benefits))

head(s4.top.benefits, n=10)

Department          avg_retirement avg_health_dental avg_other_benefits
1  Law Library         36376.73       17529.18          10882.022         
2  LLB Law Library     34321.37       11322.73          15345.517         
3  City Attorney       26699.52       13131.92           9520.186         
4  Public Defender     22706.35       13472.78           8555.060         
5  Sheriff             27219.91       13614.85           3377.200         
6  CAT City Attorney   23712.49       11016.67           8401.298         
7  Retirement Services 22004.62       13471.70           7570.591         
8  District Attorney   22359.92       12155.40           7526.675         
9  PDR Public Defender 20832.41       10944.68           7523.766         
10 Controller          19007.48       12451.41           7767.492         
   total_benefits
1  64787.93      
2  60989.62      
3  49351.62      
4  44734.19      
5  44211.96      
6  43130.46      
7  43046.91      
8  42042.00      
9  39300.86      
10 39226.38

**Comment:** This subset breaksdown benefits across different departments. The newspaper can show the differences in the variations of benefit packages across departments.

### 3.5 Union Employee Count

In [14]:
# Subset 5
s5.count.union <- df %>%
    group_by(Union) %>%
    summarise(count = n()) %>%
    arrange(desc(count))

head(s5.count.union, n=10)

Union                                              count 
1  SEIU, Local 1021, Misc                             115502
2  SEIU - Miscellaneous, Local 1021                   108284
3  Prof & Tech Engineers - Miscellaneous, Local 21     48814
4  Prof & Tech Eng, Local 21                           43424
5  SEIU - Staff and Per Diem Nurses, Local 1021        28721
6  Police Officers' Association                        24832
7  SEIU, Local 1021, RN                                24010
8  Transport Workers - Transit Operators, Local 250-A  23560
9  SEIU - Health Workers, Local 1021                   21289
10 Laborers, Local 261                                 20158

**Comment:** According to the [U.S Bureau of Labor Statistics](https://www.bls.gov/regions/west/news-release/unionmembership_california.htm), 16.2% of the workforce in California were unionized in 2020. This subset shows the number of individuals in each union in SF which can be of interest to readers.

## 4. Single Vector

In [15]:
# Create a vector of the top 5 highest paid jobs from subset 2
jobs.vector <- c("Chief Investment Officer","Physician Administrator, DPH","Managing Director ","Sup Psychiatric Physician Spec","Chief, Fire Department")
names(jobs.vector) <- c(1:5)

jobs.vector

1                                2 
      "Chief Investment Officer"   "Physician Administrator, DPH" 
                               3                                4 
            "Managing Director " "Sup Psychiatric Physician Spec" 
                               5 
        "Chief, Fire Department"

**Comment:**
1. A Chief Investment Officer (CIO) manages the investment portfolio of a company and plays a crucial role in shaping financial strategy. They make long-term investments using the company's capital to maximize returns.
2. A Physician Administrator, DPH is a healthcare exceutrive who works within the Department of Public Health. They manage budgets, compliance and healthcare processes and are responsible for medical and administrative operations.
3. A Managing Director is a senior executive in a specific division of a company who is responsible for executing strategies for growth. They often report directly to the company's CEO and/or board of directors.
4. A Supervising Pyschiatric Physician Specialist is a medical professional in a pyschiatric/mental health facility. They manage clinical programs, ensures healthcare compliance, and directs healthcare staff and services to patients.
5. A Chief of a Fire Department is the highest ranking officer in the fire department. They are responsible for managing budget and personnel and decides on strategic planning of emergency reponse services.

## 5. Five Different Lists

In [16]:
# Create 5 lists of Total.Salary for the most recent 5 years from 2016 to 2020
list.2016 <- list(df$Total.Salary[df$Year == 2016])
list.2017 <- list(df$Total.Salary[df$Year == 2017])
list.2018 <- list(df$Total.Salary[df$Year == 2018])
list.2019 <- list(df$Total.Salary[df$Year == 2019])
list.2020 <- list(df$Total.Salary[df$Year == 2020])

In [17]:
# Print first 5 values in each list
head(list.2016[[1]], n=3)
head(list.2017[[1]], n=3)
head(list.2018[[1]], n=3)
head(list.2019[[1]], n=3)
head(list.2020[[1]], n=3)

[1] 75147.32 79836.00 17806.60

[1]  77496.07 105997.00  97929.54

[1]  81096.04 115053.11 107355.32

[1]  91332.01 123241.00 115715.39

[1]  95416.01 125989.00 118150.24

**Comment:** Each list represents the total salary figures for employees in the years 2016-2020. These lists provide valuable insights into compensation from a business perspective. By comparing each year's data to previous years, the newspaper can track year-to-year changes in salaries. Depending on whether salaries increase or decrease per year, salary trends over time can provide crucial insights into understanding workforce dynamics and economic factors.

## 6. Conditionals

In [18]:
# Find quartiles from the subset 1: s1.avg.salaries
quartiles <- quantile(s1.avg.salaries$avg_salaries, probs = c(0.25,0.5,0.75))

# Store quartiles in variables 
q1 <- quartiles[1]
median <- quartiles[2]
q3 <- quartiles[3]

# Print quartiles
q1
median
q3

25% 
60512.72

50% 
75552.32

75% 
99875.44

In [19]:
# Conditional 1: Filter records on less than or equal to Q1 (25th percentile)
s1.avg.salaries %>%
    filter(avg_salaries <= q1) %>%
    head(5)

job_family                   avg_salaries
1 Revenue                      60250.26    
2 Pub Relations & Spec Assts   58778.88    
3 Protection & Apprehension    55053.15    
4 Semi-Skilled & General Labor 53579.28    
5 SF Redevelopment Agency      53426.03

**Comment:** We are filtering records from subset 1 which contains average total salaries by job family. The goal is to identify employees whose averages are below or equal to the 25th percentile of average salaries. This conditional helps identify employee salaries that are among the bottom 25% of job families.

In [20]:
# Conditional 2: Filter records on less than or equal to Q2/Median (50th percentile)
s1.avg.salaries %>%
    filter(avg_salaries <= median) %>%
    head(5)

job_family                    avg_salaries
1 Budget, Admn & Stats Analysis 75552.32    
2 Community Development         75343.62    
3 Street Transit                74214.37    
4 Appraisal & Taxation          72434.28    
5 Human Services                70640.62

**Comment:** We filter records from subset 1 that are less than or equal to the median total salary which is the 50% percentile. This conditional helps identify employees whose salaries are below the median.

In [21]:
# Conditional 3: Filter records on less than or equal to Q3 (75th percentile)
s1.avg.salaries %>%
    filter(avg_salaries <= q3) %>%
    head(5)

job_family                     avg_salaries
1 Professional Engineering       99590.61    
2 Emergency Coordination         97256.02    
3 Health & Sanitation Inspection 96808.90    
4 Journeyman Trade               94592.40    
5 Property Administration        92581.28

**Comment:** We filter records from subset 1 that are less than or equal to the 75th percentile which represents the upper quartile. This conditional identifies employees whose salaries are in the top 25%.

## 7. Loops

In [22]:
# Loop 1
# Get unique values of years
years <- unique(df$Year)

# Create empty vector to store average overtime pay for each year
avg_ot_yr <- numeric(length(years))

# Loop thru each year and calculate the average overtime pay
for (i in c(1:8)) {
    year_data <- subset(df, Year == years[i])
    avg_ot <- mean(year_data$Overtime, na.rm = TRUE)
    avg_ot_yr[i] <- avg_ot
}

# Create dataframe and store results
average_overtime_data <- data.frame(year = years, avg_overtime = avg_ot_yr)

# Print dataframe
print(average_overtime_data)

  year avg_overtime
1 2013     4249.039
2 2014     4305.350
3 2015     4411.050
4 2016     4928.004
5 2017     5675.529
6 2018     6490.329
7 2019     7076.575
8 2020     7306.643


**Comment:** This loop anlayze the average overtime pay per year by iterating thru each year in the dataset and calculating the average overtime pay each year. The results are stored in a dataframe.

In [23]:
# Loop 2
# Get unique values for years and departments
years <- unique(df$Year)
departments <- unique(df$Department)

# Create empty data frame to store results
employee_count_data <- data.frame(Year = numeric(0), Department = character(0), Employee_Count = numeric(0))

# Loop through each year and department and count # of employees
for (yr in years) {
      for (dept in departments) {
        employee_count <- sum(df$Year == yr & df$Department == dept)
        employee_count_data <- rbind(employee_count_data, data.frame(year = yr, department = dept, count = employee_count))
}}

# Sort by count and print
employee_count_data %>%
    arrange(desc(count)) %>%
    head(10)

year department                    count
1  2019 Public Health                 19788
2  2018 Public Health                 19078
3  2016 DPH Public Health             18955
4  2017 Public Health                 18833
5  2015 DPH Public Health             18349
6  2014 DPH Public Health             17966
7  2013 DPH Public Health             17823
8  2019 Municipal Transportation Agcy 14037
9  2016 MTA Municipal Transprtn Agncy 13246
10 2015 MTA Municipal Transprtn Agncy 13043

**Comment:** This loop counts the number of employees in each department for each year that spans our data and outputs a dataframe that is sorted by the count of employees.

In [24]:
# Loop 3
# Choose year input: 2013-2020
year_input <- 2017

# Create variables to count employees in union or not in union
in_union_count <- 0
not_in_union_count <- 0

# Loop through each row of the dataset for the input year and count
for (i in 1:nrow(df)) {
    if (df$Year[i] == year_input) {
        if (!is.na(df$Union[i]) & (df$Union[i] != "")) {
            # Employee in a union
            in_union_count <- in_union_count + 1
        } else {
            # Employee not in a union
            not_in_union_count <- not_in_union_count + 1
}}}

# Print the results
cat("Number of Employees in a union in", year_input,":", in_union_count, "\n")
cat("Number of Employees not in a union in", year_input,":",  not_in_union_count, "\n")

Number of Employees in a union in 2017 : 129821 
Number of Employees not in a union in 2017 : 172 


**Comment:** This loop counts how many employees are in a union vs. not in a union given a year input. The newspaper can adjust the year input and outputs the corresponding values.

## 8. Summary

This report provides an overview of the dataset depicting the employees employed by the city of San Francisco. The dataset contains a total of 683,277 rows and 22 columns and spans from 2013-2020. Among these records, there a total of 309,750 unique employees. Our anlysis shows insights within Job Families and Salaries such as "Administrative & Mgmt (Unrep)" with average salaries of \\$238,194 and "Fire Services" with average salaries of $142,523. This report also identifies the highest-paying positions such as the "Chief Infvestment Officer" with an average total compensation of \\$494,296. We also cover the benefit variations between various department as well as union participation.

## 9. Recommendations

We recommend that the newspaper uses key findings in this report as a guide to publish series of articles that takes a deep dive into the cost of living. We also recommend that the newspaper looks into other areas of analysis that were not covered in this report. This dataset includes a fairly large dataset with 22 features and not all features were used in this report. However, the certain key areas that are highlighted in this report are salary disparities between job families, profiles of high paying jobs, benefit analysis between departments, union membership insights, and yearly workforce trends. We recommend that the newspaper also creates visualization that may aid in capturing readers' attention that can help illustrate the cost of living analysis.